### DISCUSSION TITLE
[help] Monitor CPU and Memory Workload

### DISCUSSION THREAD

**Discussion by rsangole** [ropensci/targets#1256] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hello,

I've recently ported a large workload from a script to a targets pipeline. I need to monitor CPU and memory utilization for this process, to do some optimization work.

In the script world, I simply setup a monitoring shell script like so:

`monitor.sh`:

```sh 
#!/bin/bash

outfile=$1
interval=$2
process=$3

echo "ts,proc,cpu_pc,rss,vsz" > $outfile
while true; do
    echo -n "$(date +%s),$process," >> $outfile
    ps -p $process -o %cpu=,rss=,vsz= | awk '{print $1","$2","$3}' >> $outfile
    sleep $interval
done
```

And then used `{processx}` to fire the script, run the large compute, and then kill the process.

```r
library(processx)

p <- process$new(
    command = "monitor.sh",
    args = c(
        out_file_name,
        1,
        Sys.getpid() # <<--- Process ID of the current R process
    )
)

very_large_compute()

p$kill()
```

This worked great since the PID of the current R process was the PID I had to monitor.

How can I extend this in the targets framework, with multiple crew workers working through dynamic branches?

Tbh, I don't need need to monitor each individual worker or branch... would love to get an "overall" targets process CPU and Mem consumption over time.

Any ideas?

**Comment by wlandau** [comment]
This is outside the scope of `targets` itself because `targets` only operates within the R sessions (local and worker R processes) where it runs, whereas monitoring CPU/RAM requires tools that run outside these R sessions. If you are okay with interactive tools, then `htop` is convenient. Otherwise the `ps` package can help. If everything is running on your local computer (as opposed to a cluster or cloud service) then you can identify all the relevant local processes by searching the process commands for mentions of "targets::", "crew::", and "mirai::".

**Reply by rsangole** [reply]
Thanks Will. I think that's the approach I'll take. Was just wondering of `targets` has an inbuilt function which tracks (and thus can return to me) all the process IDs launched by `crew`.
